### ISSUE TITLE
Green Go Communication Version 5 

### ISSUE THREAD

**Issue by nickjfw94** [bitfocus/companion-module-requests#722]
Hello.
I have had a search and couldn't see this. 

Green Go Comms that are a widely used communication product has now enables Scripting & API calls in the new version 5. V5 is a massive update and for the first time has allowed API integration. It looks like OSC is supported so I might take a look at this too.

Details can be found here: https://manual.greengoconnect.com/en/scripting/api/

**Comment by PhilGDS** [comment]
Seconded this. We're in a studio running beltpacks in gallery positions. Being able to do more complex commands simply via StreamDeck/OSC would be perfect.

**Comment by krisdahl** [comment]
Just to update, the Green Go scripting language doesn't have like a REST API or OSC that we can hit directly.  But the Green Go scripting language does support listening to and firing OSC commands. 

So to implement this we'd need to have a listener script running on Green Go that we'd then fire commands to from Companion.

Do either of you  @PhilGDS or @nickjfw94 have a use case of what you'd like to be able to do?  There are a lot of API calls that could be made--but it would be nice to focus on something real world to start with rather than trying to expose the entire API via a Green Go Script that is then implemented in Companion.

**Comment by DaveTech78** [comment]
I use greengo as well, I'd like to be able to latch/unlatch listens and have PTT functions.

**Comment by PhilGDS** [comment]
Thanks Krisdahl! 

As we're using beltpacks, we have Comms channels available at face value, or a flexlist that requires a bunch of scrolling.

Examples that would be good for us include;
- Direct comms to particular user (done by ID). We have Users define with pack presets here, per studio i.e. Studio 1 Technical Operator A, and a Studio 2 Technical Operator A. S1 Tech op is user ID 4, and S2 Tech Op is user ID 9. We currently run 3 studios with 6/7 users per studio. 
- PPT and Latch - option to do with this Companion native latch. I think there's an OSC command to do that action itself, which is an extra step. 
- Audio Control, e.g. Gain. Some of our users have softer voices, and other louder, so giving them more button control would be easier. Potentially even saving a script to load their personal settings. 
- Volume control - replacing the use of the encoder wheels for main volume and individual channels. 
- Feedback to the online status of a pack!

The Goal for this would be looking to hide the beltpacks/tidy them away, as they're just floating on our gallery desks currently/jury rigged with a strap to the side.

**Comment by PhilGDS** [comment]
@krisdahl - Having a think of this again, just for my understanding, would the idea be to upload a singular script to the Green-Go device that is set to listen, and all commands would be built as a module in Companion that automatically puts together the OSC commands (as opposed to making a generic OSC module and writing all the integers and strings)?

**Comment by krisdahl** [comment]
> @krisdahl - Having a think of this again, just for my understanding, would the idea be to upload a singular script to the Green-Go device that is set to listen, and all commands would be built as a module in Companion that automatically puts together the OSC commands (as opposed to making a generic OSC module and writing all the integers and strings)?

Yes, that'd be ideal to have a script with all the core API calls that you run on the Green Go device.  Then a plugin on Companion that is designed to hit those (via OSC) rather than a generic OSC command.

I'll experiment with it this week maybe and see if the latency is good enough to use OSC commands + stream deck to use as expansion panel.

**Comment by lampensau** [comment]
Hey there,

I've just been made aware of this issue/request by @PhilGDS.

I've been working on a Companion Module for Green-GO v5 over the last couple of days. It is pretty much complete and can be reviewed here: [companion-module-greengo-intercom](https://github.com/bitfocus/companion-module-greengo-intercom). I think I will keep testing this until Monday and then request a review.

The linked repository also includes an `osc-remote.gg5t` script file that needs to run on the Green-GO device to be controlled. Otherwise, the Companion module will not function.

**Comment by ryanfelipe8** [comment]
Hello, I do some events with greengo where I need to talk to more than 20 people each at a time, but I saw that the version available in the companion only allows me to create shortcuts for the first 6 channels, we would be able to increase the number of channels for 32?

**Comment by lampensau** [comment]
Hey @ryanfelipe8,

Unfortunately, this is currently not as easy as I'd like it to be. Therefore it is currently not possible to adapt the existing Companion module. 

However, it would be possible to use many more channels in a custom implementation where unneeded features like (for example) cue or call signals. Such implementation is possible using the existing OSC module for Companion and using a custom Green-GO script running on the device.

I'm unfortunately currently very short on time. However, please feel free to ask for any assistance on a custom implementation. Additionally, feel free to ask for help on the [Green-GO user forums](https://greengoconnect.com).

**Comment by ryanfelipe8** [comment]
Thank you very much for your response

I will need this customization, we can agree on a price for this task, if you have some free time, send me an email at: lincoln@workflow.com.co
